crates/hir/src/lib.rs

```diff
@@ -4926,28 +4926,49 @@ impl Type {
         traits_in_scope: &FxHashSet<TraitId>,
         with_local_impls: Option<Module>,
         name: Option<&Name>,
-        mut callback: impl FnMut(Function) -> Option<T>,
+        callback: impl FnMut(Function) -> Option<T>,
     ) -> Option<T> {
+        struct Callback<T, F> {
+            f: F,
+            slot: Option<T>,
+        }
+        impl<T, F> MethodCandidateCallback for &'_ mut Callback<T, F>
```

Comment:
```suggestion
        impl<T, F> MethodCandidateCallback for Callback<T, F>
```
shouldn't this work given the methods take `&mut self` (would avoid a double indirection)

Ah right you want to keep the value to access slot later. Could make `slot` be an `&mut Option<T>` instead then I guess. Though that likely doesn't matter